rabbit-escape-engine/src/rabbitescape/engine/util/Util.java

```diff
@@ -631,13 +663,23 @@ protected String nextOrNull()
         return streamLines( name, res );
     }
 
-    public static <T> T[] concat( T[] left, T[] right )
-    {
-        return list(
-            chain(
-                Arrays.asList( left ), Arrays.asList( right )
-            )
-        ).toArray( left );
+    @SafeVarargs
+    public static <T> T[] concat(T[]... arrays) {
+        int totalLen = 0;
+        for (T[] arr: arrays)
+        {
+            totalLen += arr.length;
+        }
+        @SuppressWarnings( "unchecked" )
+        T[] all = (T[])Array.newInstance(
+            arrays.getClass().getComponentType().getComponentType(), totalLen);
+        int copied = 0;
+        for (T[] arr: arrays)
+        {
+            System.arraycopy(arr, 0, all, copied, arr.length);
+            copied += arr.length;
+        }
+        return all;
```

Comment:
OK. I will have another go at this. I understand the objection to the array cast.

This also removes the CCSA attribution pain.

I was wondering about calling the 2 arg concat recursively to do varargs.